File: test/lan_html.ml
open Kanon_kernel
module E = Kanon_surface.Elab
module L = Kanon_surface.Lower
module T = Lanyard_rust.Text_ops
module I = Lanyard_rust.Interp
module V = Lanyard_rust.Run_value
module H = Lanyard_rust.Run_http
module S = Lanyard_rust.Run_store
module C = Lanyard_target.Target_generated
let ( let* ) = Result.bind
let base = "mu Bytes : Type 0 := | bytesNil : Bytes | bytesCons (head : Nat) (tail : Bytes) : Bytes\n"
let fail message = Error (Error.Mismatch message)
let contains text needle = String.to_seq text |> Seq.mapi (fun index _character -> index)
  |> Seq.exists (fun index -> String.starts_with ~prefix:needle
       (String.to_seq text |> Seq.drop index |> String.of_seq))
let equal expected result = let* actual = result in
  if actual = expected then Ok () else fail "unexpected HTML result"
let refuses needle result = Result.fold ~ok:(fun _value -> fail "expected refusal")
  ~error:(fun error -> if contains (Error.to_string error) needle then Ok () else Error error) result
let result_type schema = if String.equal schema "Html.text" then "Bytes" else "Response"
let checked schema = E.check_lanyard (base ^ "def main : Bytes -> " ^ result_type schema ^ " := " ^ schema ^ " Bytes")
let call schema value =
  let* checked = checked schema in let* context = I.prepare checked in
  I.call context 0 { I.steps = I.default_steps; store = S.empty } "main" [value]
  |> Result.map fst
let escape value = let* value = call "Html.text" (V.of_text "Bytes" value) in
  V.text ~what:"HTML" "Bytes" value
let render value = let* value = call "Response.html" (V.of_text "Bytes" value) in H.response value
let wire body = "HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=utf-8\r\nContent-Length: "
  ^ string_of_int (String.length body) ^ "\r\n\r\n" ^ body
let evaluate source =
  let* checked = E.check_lanyard (base ^ source) in
  let* value, _text = I.run checked in H.response value
let metadata schema change_row change_entry =
  let* checked = checked schema in let* specialized = L.specialize checked in
  let* instances = L.text_operations specialized in
  let* operations = T.catalog specialized.specialized instances in
  let* operation = List.find_opt (fun (text : T.t) -> text.row.schema = schema) operations
    |> Option.to_result ~none:(Error.Unbound schema) in
  let entries = List.map (fun (entry : C.entry) ->
    if entry.name = schema then change_entry entry else entry) C.entries in
  T.foreign_call entries operations (change_row operation.row)
let malformed = V.Tag (Erase.mu_tid "Bytes", 1, [V.Nat (Bignum.of_int 256); V.of_text "Bytes" ""])
let cases = [
  "escape delimiters", equal "&lt;a&gt;&amp;&lt;/a&gt;" (escape "<a>&</a>");
  "escape existing entities", equal "&amp;lt;&amp;amp;" (escape "&lt;&amp;");
  "escape quotes retained", equal "\"'\096=/" (escape "\"'\096=/");
  "escape empty", equal "" (escape "");
  "escape Unicode", equal "caf\195\169 \230\151\165\240\159\152\128" (escape "caf\195\169 \230\151\165\240\159\152\128");
  "escape controls", equal "\000\r\n" (escape "\000\r\n");
  "escape public UTF-8 guard", refuses "UTF-8" (T.html_text "\255");
  "html raw markup", equal (wire "<h1>hello</h1>") (render "<h1>hello</h1>");
  "html empty body", equal (wire "") (render "");
  "html byte length", equal (wire "caf\195\169") (render "caf\195\169");
  "html body controls", equal (wire "\000\r\n\r\nX-Header:yes") (render "\000\r\n\r\nX-Header:yes");
  "html runtime UTF-8 guard", refuses "UTF-8" (H.response (V.Response_html "\255"));
  "composed escaping", equal (wire "&lt;script&gt;&amp;&lt;/script&gt;")
    (evaluate "def main : Response := Response.html Bytes (Html.text Bytes b\"<script>&</script>\")");
  "captured helper", equal (wire "&lt;title&gt;")
    (evaluate ("def main : Response := let title : Bytes := b\"<title>\" in "
     ^ "let escaped : prod () -> Bytes := fun (u : prod ()) => Html.text Bytes title in Response.html Bytes (escaped ())"));
  "text response retained", equal "HTTP/1.1 200 OK\r\nContent-Type: text/plain; charset=utf-8\r\nContent-Length: 3\r\n\r\n<b>"
    (evaluate "def main : Response := Response.text Bytes b\"<b>\"");
  "unused escape fails", refuses "UTF-8" (evaluate
    "def main : Response := let ignored : Bytes := Html.text Bytes (bytesCons 255 bytesNil) in Response.html Bytes b\"ok\"");
  "unused response fails", refuses "UTF-8" (evaluate
    "def main : Response := let ignored : Response := Response.html Bytes (bytesCons 255 bytesNil) in Response.html Bytes b\"ok\"");
] @ List.concat_map (fun schema -> [
  schema ^ " byte range", refuses "0..255" (call schema malformed);
  schema ^ " malformed UTF-8", refuses "UTF-8" (call schema (V.of_text "Bytes" "\255"));
  schema ^ " truncated UTF-8", refuses "UTF-8" (call schema (V.of_text "Bytes" "\226\130"));
  schema ^ " wrong family", refuses "byte list" (call schema (V.of_text "Other" "hello"));
  schema ^ " unsupported layout", refuses "byte list"
    (let* checked = E.check_lanyard ("def main : "
      ^ (if String.equal schema "Html.text" then "Nat" else "Response")
      ^ " := " ^ schema ^ " Nat 1") in I.run checked);
  schema ^ " open type", refuses "closed"
    (let* checked = E.check_lanyard ("def main : (0 A : Type 0) -> A -> "
      ^ (if String.equal schema "Html.text" then "A" else "Response")
      ^ " := fun (0 A : Type 0) (text : A) => " ^ schema ^ " A text") in I.prepare checked);
  schema ^ " metadata effects", refuses "metadata" (metadata schema Fun.id (fun row -> { row with effects = ["topcoat::Error"] }));
  schema ^ " metadata arity", refuses "metadata" (metadata schema (fun row -> { row with arity = 2 }) Fun.id);
  schema ^ " metadata quantities", refuses "metadata" (metadata schema Fun.id (fun row -> { row with quantities = [C.Zero; C.One] }));
  schema ^ " metadata kind", refuses "kind" (metadata schema Fun.id (fun row -> { row with kind = C.Constant }));
  schema ^ " metadata type", refuses "metadata" (metadata schema Fun.id (fun row -> { row with kernel_type = T.specification T.Is_empty }));
  schema ^ " metadata type argument", refuses "metadata" (metadata schema (fun row -> { row with type_arguments = [] }) Fun.id);
  schema ^ " metadata print rule", refuses "metadata" (metadata schema Fun.id (fun row -> { row with print_rule = "#{other}" }));
]) ["Html.text"; "Response.html"]
let () =
  let failed = List.filter_map (fun (name, result) ->
    Result.fold ~ok:(fun () -> None) ~error:(fun error -> Some (name ^ ": " ^ Error.to_string error)) result) cases in
  List.iter prerr_endline failed;
  if List.is_empty failed then Printf.printf "LAN-HTML OK checks=%d\n" (List.length cases) else exit 1

File: test/lan_html.py
"""Check HTML wire output, escaped model fields and checked native emission."""
from pathlib import Path
import subprocess
import tempfile
import unittest

ROOT = Path(__file__).resolve().parent.parent
DRIVER = ROOT / "_build/default/bin/lanyard.exe"
BASE = "mu Bytes : Type 0 := | bytesNil : Bytes | bytesCons (head : Nat) (tail : Bytes) : Bytes\n"


def run(*args):
    return subprocess.run([str(DRIVER), *map(str, args)], cwd=ROOT, capture_output=True, timeout=60)


def response(body):
    return (b"HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=utf-8\r\nContent-Length: "
            + str(len(body)).encode() + b"\r\n\r\n" + body)


class HtmlCli(unittest.TestCase):
    def test_markup_request(self):
        result = run("run", "--request", "/hello", "test/fixtures/html.lan")
        self.assertEqual((result.returncode, result.stdout, result.stderr),
                         (0, response(b"<h1>Hello from Lanyard!</h1>"), b""))

    def test_escaped_stored_titles(self):
        cases = [
            ("title=++write+tests++", b"write tests"),
            ("title=caf%C3%A9", "café".encode()),
            ("title=", b""),
            ("title=%3Cscript%3Ealert%281%29%3C%2Fscript%3E", b"&lt;script&gt;alert(1)&lt;/script&gt;"),
            ("title=%26lt%3B%26", b"&amp;lt;&amp;"),
            ("title=%22%27", b"\"'"),
            ("title=a%00b", b"a\0b"),
            ("title=a%0D%0A%0D%0AX-Header%3Ayes", b"a\r\n\r\nX-Header:yes"),
        ]
        for form, expected in cases:
            with self.subTest(form=form):
                result = run("run", "--form", form, "--request", "/todos", "test/fixtures/html-form.lan")
                self.assertEqual((result.returncode, result.stdout, result.stderr), (0, response(expected), b""))

    def test_no_partial_output(self):
        for form, status, message in [("", 1, b"missing field"), ("title=%", 64, b"usage:"),
                                      ("title=x&title=y", 64, b"usage:"), ("other=x", 1, b"missing field")]:
            with self.subTest(form=form):
                result = run("run", "--request", "/", "--form", form, "test/fixtures/html-form.lan")
                self.assertEqual((result.returncode, result.stdout), (status, b""))
                self.assertIn(message, result.stderr)
        result = run("run", "--steps", "1", "--request", "/", "test/fixtures/html.lan")
        self.assertEqual((result.returncode, result.stdout), (1, b""))
        self.assertIn(b"step limit", result.stderr)

    def test_invalid_bodies(self):
        with tempfile.TemporaryDirectory(prefix="lanyard-html-") as directory:
            source = Path(directory) / "invalid.lan"
            for escaped in (False, True):
                for byte, message in [(256, b"0..255"), (255, b"UTF-8")]:
                    with self.subTest(escaped=escaped, byte=byte):
                        argument = f"(bytesCons {byte} bytesNil)"
                        if escaped:
                            argument = f"(Html.text Bytes {argument})"
                        source.write_text(BASE + "def main : Cx -> Uri -> Response := fun (cx : Cx) (uri : Uri) => "
                                          f"Response.html Bytes {argument}\n")
                        result = run("run", "--request", "/", source)
                        self.assertEqual((result.returncode, result.stdout), (1, b""))
                        self.assertIn(message, result.stderr)

    def test_emission(self):
        for name in ("html", "html-form"):
            result = run("emit", "--target", f"test/fixtures/{name}.lan")
            self.assertEqual((result.returncode, result.stderr), (0, b""))
            self.assertIn(b"topcoat::router::Body::from", result.stdout)
            self.assertIn(b"text/html; charset=utf-8", result.stdout)
        with tempfile.TemporaryDirectory(prefix="lanyard-html-crate-") as directory:
            source = Path(directory) / "main.lan"
            source.write_text(BASE + 'def main : Response := Response.html Bytes (Html.text Bytes b"<hello>")\n')
            destination = Path(directory) / "crate"
            result = run("emit", "--crate", destination, source)
            self.assertEqual((result.returncode, result.stderr), (0, b""))
            emitted = (destination / "src/main.rs").read_text()
            self.assertIn("text/html; charset=utf-8", emitted)
            self.assertIn(".replace('&', \"&amp;\")", emitted)

    def test_foreign_ownership_limits(self):
        cases = [
            ('def main : Response := let value : Response := Response.html Bytes b"x" in value',
             b"owned copy of a shared foreign value"),
            ('def main : prod (Response, Response) := let value : Response := Response.html Bytes b"x" in tuple (value, value)',
             b"foreign aggregate layout"),
        ]
        with tempfile.TemporaryDirectory(prefix="lanyard-html-layout-") as directory:
            source = Path(directory) / "main.lan"
            for declaration, message in cases:
                with self.subTest(declaration=declaration):
                    source.write_text(BASE + declaration + "\n")
                    result = run("emit", "--target", source)
                    self.assertNotEqual(result.returncode, 0)
                    self.assertEqual(result.stdout, b"")
                    self.assertIn(message, result.stderr)


if __name__ == "__main__":
    unittest.main()
